src/mimemail.erl

```diff
@@ -916,28 +916,195 @@ has_lines_over_998(Bin, {FoundAt, 2}, _, Pattern) ->
+    [<<Name/binary, $=, Plain/binary>>];
+encode_parameter(plain, Name, Index, Value, EncLen, Acc) ->
+    Prefix = <<Name/binary, $*, (integer_to_binary(Index))/binary>>,
+    {Plain, More} = encode_param_value(plain, Value, 75 - byte_size(Prefix)),
+    encode_parameter(plain, Name, Index + 1, More, EncLen, [<<Prefix/binary, $=, Plain/binary>> | Acc]).
+
+%% Encode a parameter value according to the method
+%% given as the first argument.
+-spec encode_param_value(Method :: (plain | quote | encode), Value :: binary(), Len :: integer()) ->
+    {Chunk :: binary(), Rest :: binary()}.
+encode_param_value(plain, Value, Len) ->
+    %% No encoding necessary, return (part of) the
+    %% value as-is.
+    Len1 = max(Len, 1),
+    case Value of
+        <<Part:Len1/bytes, More/binary>> ->
+            {Part, More};
+        _ ->
+            {Value, <<>>}
+    end;
+encode_param_value(quote, Value, Len) ->
+    %% See encode_param_value_quote/3
+    encode_param_value_quote(Value, Len, <<>>);
+encode_param_value(encode, Value, Len) ->
+    %% See encode_param_value_encode/3
+    encode_param_value_encode(Value, Len, <<>>).
+
+%% Encode (part of) a parameter value by quoting.
+%% " and \ are escaped by preceding it with a \,
+%% everything else remains as-is.
+encode_param_value_quote(<<>>, _Len, Acc) ->
+    {Acc, <<>>};
+encode_param_value_quote(All = <<C, More/binary>>, Len, Acc) ->
+    case C =:= $" orelse C =:= $\\ of
+        true when Len >= 2; Acc =:= <<>> ->
+            encode_param_value_quote(More, Len - 2, <<Acc/binary, $\\, C>>);
+        false when Len >= 1; Acc =:= <<>> ->
+            encode_param_value_quote(More, Len - 1, <<Acc/binary, C>>);
+        _ ->
+            {Acc, All}
     end.
 
-% See also: http://www.ietf.org/rfc/rfc2045.txt section 5.1
-escape_tspecial(<<>>, false, _Acc) ->
-    false;
-escape_tspecial(<<>>, IsSpecial, Acc) ->
-    {IsSpecial, Acc};
-escape_tspecial(<<C, Rest/binary>>, _IsSpecial, Acc) when C =:= $" ->
-    escape_tspecial(Rest, true, <<Acc/binary, $\\, $">>);
-escape_tspecial(<<C, Rest/binary>>, _IsSpecial, Acc) when C =:= $\\ ->
-    escape_tspecial(Rest, true, <<Acc/binary, $\\, $\\>>);
-escape_tspecial(<<C, Rest/binary>>, _IsSpecial, Acc) when
+%% Encode (part of) a parameter value according to RFC2231.
+%% Tspecials, CTLs, spaces, *, ', % and the bytes of
+%% multi-byte UTF-8 characters are encoded as a % followed
+%% by the hex representation of the byte values.
+%% Everything else remains as-is.
+encode_param_value_encode(<<>>, _Len, Acc) ->
+    {Acc, <<>>};
+encode_param_value_encode(All = <<C/utf8, More/binary>>, Len, Acc) when
+    C =< 16#1F;
+    C =:= 16#7F;
+    C =:= $(;
+    C =:= $);
+    C =:= $<;
+    C =:= $>;
+    C =:= $@;
+    C =:= $,;
+    C =:= $;;
+    C =:= $:;
+    C =:= $/;
+    C =:= $[;
+    C =:= $];
+    C =:= $?;
+    C =:= $=;
+    C =:= $\s;
+    C =:= $*;
+    C =:= $';
+    C =:= $%
+->
+    BinC = <<C/utf8>>,
+    EncSize = 3 * byte_size(BinC),
```

Comment:
Just wondering. Given the allowed values in the when-clause, isn't the byte size of BinC always 1?

Reply:
Hm, you're right, and also the subsequent handling of the character is "off". I mean, it does the right thing, but overly complicated. I'm not quite sure what I was thinking back then I have to admit, given that I did that almost a year ago :sweat_smile: I probably restructured things at some point and forgot to pull it all straight. I'll look into it again, but probably not today. So don't merge this yet :wink:

Reply:
Done :)

Reply:
That looks a lot simpler :-)